calicoctl/commands/diags.go

```diff
@@ -0,0 +1,188 @@
+
+import (
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/docopt/docopt-go"
+)
+
+type regPair struct {
+	reg     string
+	replace string
+}
+
+// Diags function collects diagnostic information and logs
+func Diags(args []string) error {
+	var err error
+	doc := `Usage: 
+            calicoctl diags [--log-dir=<LOG_DIR>]
+
+    Description:
+         Save diagnostic information
+
+    Options:
+         --log-dir=<LOG_DIR>  The directory for logs [default: /var/log/calico]`
+
+	arguments, err := docopt.Parse(doc, args, true, "calicoctl", false, false)
+	if err != nil {
+		return err
+	}
+
+	runDiags(arguments["--log-dir"].(string))
+
+	return nil
+}
+
+func runDiags(logDir string) {
+	cmds := []struct {
+		info string
+		cmd  string
+	}{
+		{"", "date"},
+		{"", "hostname"},
+		{"Dumping netstat", "netstat --all --numeric"},
+		{"Dumping routes (IPv4)", "ip -4 route"},
+		{"Dumping routes (IPv6)", "ip -6 route"},
+		{"Dumping interface info (IPv4)", "ip -4 addr"},
+		{"Dumping interface info (IPv6)", "ip -6 addr"},
+		{"Dumping iptables (IPv4)", "iptables-save"},
+		{"Dumping iptables (IPv6)", "ip6tables-save"},
+		{"Dumping ipsets", "ipset list"},
+		{"Dumping ipsets (container)", "docker run --privileged --net=host calico/node ipset list"},
+		{"Copying journal for calico-node.service", "journalctl -u calico-node.service --no-pager"},
+	}
+
+	// Make sure the command is run with super user priviladges
+	if os.Getuid() != 0 {
+		fmt.Println("Need super user privilages: Operation not permitted")
+		os.Exit(1)
+	}
+
+	fmt.Println("Collecting diagnostics")
+
+	// Create a temp directory in /tmp
+	tmpDir, err := ioutil.TempDir("/tmp", "tmp")
+	if err != nil {
+		log.Fatalf("Error creating temp directory to dump logs: %v\n", err)
+	}
+
+	fmt.Println("Using temp dir:", tmpDir)
+	err = os.Chdir(tmpDir)
+	if err != nil {
+		log.Fatalf("Error changing directory to temp directory to dump logs: %v\n", err)
+	}
+
+	os.Mkdir("diagnostics", os.ModeDir)
+	diagsTmpDir := filepath.Join(tmpDir, "diagnostics")
+
+	for _, v := range cmds {
+		writeDiags(v.info, v.cmd, diagsTmpDir)
+	}
+
+	fmt.Println("Dumping felix stats")
+	err = exec.Command("pkill", "-SIGUSR1", "felix").Run()
+	if err != nil {
+		log.Printf("Error killing the felix process: %v\n", err)
+	}
+
+	fileInfo, err := os.Stat(logDir)
+
+	// Check if the logDir provided/default exists and is a directory
+	if fileInfo.IsDir() {
+		fmt.Println("Copying Calico logs")
+		cpCmd := exec.Command("cp", "-rf", logDir, diagsTmpDir)
+		err = cpCmd.Run()
+		if err != nil {
+			log.Fatalf("Error copying log files: %v\n", err)
+		}
+	} else {
+		fmt.Printf("No logs found in %s; skipping log copying", logDir)
+	}
+
+	// Get the current time and create a tar.gz file with the timestamp in the name
+	timeNow := time.Now()
+	y, mo, d := timeNow.Date()
+	h, m, s := timeNow.Clock()
+	tarFile := fmt.Sprintf("diags-%d%d%d_%d%d%d.tar.gz", d, mo, y, h, m, s)
```

Comment:
using RFC3399 also means our output tar file name format would be different. I did the same format DMY_HMS python version has for the filename, or should I change it to `20160913_131450`?